stage #0/module #4 OOP Fundamentals/oop_fundamentals.md

```diff
@@ -0,0 +1,777 @@
+Software objects are conceptually similar to real-world objects: they consist of state and related behavior as well. An object stores its state in fields (variables) and defines its behavior through methods (functions). Methods operate on an object's internal state and serve as the primary mechanism for communication among objects. Fiels and methods of classes are called their members. It is common practise to name class members in lowercase in compliance with ["camelCase" style](https://betterprogramming.pub/string-case-styles-camel-pascal-snake-and-kebab-case-981407998841). 
+More detailed information about objects can be found [here](https://docs.oracle.com/javase/tutorial/java/concepts/object.html).
+
+## Classes
+Class is a template, which objects are created from. In the real world, we can find many individual objects, which are of the same kind. There may be thousands of smartphones, which have the same make and model. Each this smartphone was built from the same set of blueprints and therefore contains the same components. In object-oriented terms, we say that somebody's smartphone is an instance of the class of objects known as "smartphones".
+
+Example
+```
+class Smartphone {
+
+    public Smartphone() {
+        // it is a constructor, which a special method  
+    }
+
+    boolean isScreenTurnedOn = false; // an initial state of every smartphone is a turned-off screen;
+
+    void turnOnScreen() {
+        isScreenTurnedOn = true; // if a method is called, a screen of smartphone will be turned on;
+    }
+
+    void turnOffScreen() {
+        isScreenTurnedOn = false; // if a method is called, a screen of smartphone will be turned off;
+    }
+}
+```
+
+### Declaration of Classes and its Members
+
+#### Class Declaration
+The simplest way to declare class is the following.
+```
+public class Bicycle {
+
+   // The class has a name "Bicycle";
+   
+   // This class has no state (fields) and behaviour (methods). 
+}
+```
+Here we declared an empty class body, which is an area between curly-braces ("{}").
+
+#### Fields Declaration
+In order to add state for a class, it is necessary to declare a field in a class body.
+Field declarations contain three components in the following order:
+- (Optional) Modifier, which can be public, protected, private;
+- (Required) Field Type;
+- (Required) Field Name.
+```
+public class Bicycle {
+    
+    public int speed = 0; // now a class has a state: an initial speed is "0";
+      
+    // This class still has no behaviour (methods).
+}
+```
+Modifiers allow us to control, what other classes can have access to a field (this theme will be discussed further). Type is a parameter, which describes a structure of a class state. Fields can be of primitive type (int, float, boolean, etc.) and reference type (arrays and other objects). Names should follow the same naming rules and conventions (this theme will be discussed further).
+
+#### Methods Declaration
```

Comment:
when I was studying java it was super unclear for me how functions can return anything, so maybe it would be nice to describe what and how something can be returned from a method.